draft-ietf-httpbis-message-signatures.md

```diff
@@ -382,7 +348,209 @@ This example shows a canonicalized value for the parameters of a given signature
+
+### Method {#content-request-method}
+
+The `@method` identifier refers to the HTTP method of a request message. The value of is canonicalized by taking the value of the method as a string. Note that the method name is case-sensitive as per {{!SEMANTICS=RFC7231}}, and conventionally standardized method names are uppercase US-ASCII.
+
+For example, the following request message:
+
+~~~
+POST /path?param=value HTTP/1.1
+Host: www.example.com
+~~~
+
+Would result in the following `@method` value:
+
+~~~
+"@method": POST
+~~~
+
+If used in a response message, the `@method` identifier refers to the associated value of the request that triggered the response message being signed.
+
+### Target URI {#content-target-uri}
+
+The `@target-uri` identifier refers to the target URI of a request message. The value is the full absolute target URI of the request, potentially assembled from all available parts including the authority and request target.
+
+For example, the following message sent over HTTPS:
+
+~~~
+POST /path?param=value HTTP/1.1
+Host: www.example.com
+~~~
+
+Would result in the following `@target-uri` value:
+
+~~~
+"@target-uri": https://www.example.com/path?param=value
+~~~
+
+If used in a response message, the `@target-uri` identifier refers to the associated value of the request that triggered the response message being signed.
+
+### Authority {#content-request-authority}
+
+The `@authority` identifier refers to the authority component of the target of the HTTP request message. In HTTP 1.1, this is usually conveyed using the `Host` header, while in HTTP 2 and HTTP 3 it is conveyed using the `:authority` pseudo-header. The value is the fully-qualified authority component of the request, comprised of the host and, optionally, port of the request target, as a string.
+
+For example, the following request message:
+
+~~~
+POST /path?param=value HTTP/1.1
+Host: www.example.com
+~~~
+
+Would result in the following `@authority` value:
+
+~~~
+"@authority": www.example.com
+~~~
+
+If used in a response message, the `@authority` identifier refers to the associated value of the request that triggered the response message being signed.
+
+### Scheme {#content-request-scheme}
+
+The `@scheme` identifier refers to the scheme of the target URL of the HTTP request message. The value is the scheme as a string.
+
+For example, the following request message requested over plain HTTP:
+
+~~~
+POST /path?param=value HTTP/1.1
+Host: www.example.com
+~~~
+
+Would result in the following `@scheme` value:
+
+~~~
+"@scheme": http
+~~~
+
+If used in a response message, the `@scheme` identifier refers to the associated value of the request that triggered the response message being signed.
+
+### Request Target {#content-request-target}
+
+The `@request-target` identifier refers to the full request target of the HTTP request message. The value of the request target can take different forms, depending on the type of request.
```

Comment:
You always need to be able to accurately re-create the signing input, so if that's difficult for a given API/application/deployment they wouldn't sign that part. If you have to start guessing for ANYTHING it's a bad idea to sign it. We haven't filled out the security considerations section yet but this is a key point for this protocol.

Reply:
Right, a signature is useless if the recipient can't verify it reliably, and since the elements aren't signed separately, including any element which can't be reliably reconstructed renders the entire signature meaningless.  I think you're assuming a closed system, where the server won't bother to sign something it knows some of its clients can't verify.  I'm thinking about a more open system where the server won't necessarily know the client capabilities; we should build something where the client is highly likely to be able to verify reliably.